webcrypto/algorithm.go

```diff
@@ -322,3 +116,64 @@ const (
 	// OperationIdentifierDigest represents the digest operation.
 	OperationIdentifierDigest OperationIdentifier = "digest"
 )
+
+// normalizeAlgorithm normalizes the given algorithm following the
+// algorithm described in the WebCrypto [specification].
+//
+// [specification]: https://www.w3.org/TR/WebCryptoAPI/#algorithm-normalization-normalize-an-algorithm
+func normalizeAlgorithm(rt *goja.Runtime, v goja.Value) (Algorithm, error) {
+	var algorithm Algorithm
+
+	if v.ExportType().Kind() == reflect.String {
+		algorithmString, ok := v.Export().(string)
+		if !ok {
+			return Algorithm{}, NewError(0, ImplementationError, "algorithm cannot be interpreted as a string")
+		}
+
+		algorithm.Name = algorithmString
+
+		return algorithm, nil
```

Comment:
The thing is that the specification (which I would prefer to follow to the letter whenever possible) actually explicitly states that this is how it should work.

Also in those places you linekd we again call `normalizeAlgorithm` and then it actually (should) *not* go in this if, so it will do the checks below.

Reply:
Agreed that we should closely follow the spec, in which case this doesn't seem to be doing that.

Notice that [the spec](https://www.w3.org/TR/WebCryptoAPI/#algorithm-normalization-normalize-an-algorithm) says:

> If alg is an instance of a DOMString:
>   Return the result of running the [normalize an algorithm](https://www.w3.org/TR/WebCryptoAPI/#dfn-normalize-an-algorithm) algorithm, with the alg set to a new [Algorithm](https://www.w3.org/TR/WebCryptoAPI/#dfn-Algorithm) dictionary whose [name](https://www.w3.org/TR/WebCryptoAPI/#dfn-KeyAlgorithm-name) attribute is alg, and with the op set to op.

I.e. we should return the result of `normalizeAlgorithm()` here, passing an object instead. It might help if implement the algorithm in a clean function, free of any goja-isms.

My point is that we skip the validation if it's a string, which then makes us return `NotSupportedError` in other places where we don't need to. It's a simplification that the spec'd algorithm accounts for.

BTW, if we really want to be strict about following the spec, `normalizeAlgorithm()` should take an `op` (operation) argument, and return it in the normalized version, which we currently don't do. It seems this is a simplification for behavior we don't need, but in the case of validating a string, it's clearly missing from the implementation.

Reply:
Hehe, I've spent literally hours staring at this specs in the last months, I guess I could almost recite them by heart at this point 🤝 😄 You're definitely right, I forgot to return the op indeed, which in my experience is useful when you get an error, but can't know from which promise it comes. Having a dedicated free of goja-ism function to try and handle that case is also a good idea, I'm gonna see what I can come up with. 

For the reference, i[n the initial PR, I had went through the hassle of implementing an "as close" as possible to the specs algorithm normalisation](https://github.com/grafana/xk6-webcrypto/pull/26/files#diff-6726d6a67040dc82c4a01ace327b184ce34fd868c3e16f17716b03bb3cd7a129R31). At the time, @olegbespalov argued that as long as the tests passed, it didn't matter if we implemented all the steps, especially considering that those were designed for IDL compliance; which we don't really care for in k6?

Both options are possible, and I went for a compromise 😄

Reply:
I don't mind the missing `op` handling. I agree with Oleg, and it's best to be pragmatic about this, as long as the tests pass.

But I do think we should fix the validation in `normalizeAlgorithm()` if the passed value is a string. It would be compliant with the spec, and less code overall.

Reply:
Fair enough @imiric. [I've updated the `normalizeAlgorithm` to be more accurate](https://github.com/grafana/xk6-webcrypto/pull/30/files#diff-261464395ae77be967b325f3b606881c51d29a3848d891e5bfe21561e4574df3R124), and to handle strings as-per-the-specs 🤞🏻 Let me know if that's satisfying 🙇🏻

Reply:
LGTM :slightly_smiling_face: Now you don't need to reject the promise with `NotSupportedError` in `GenerateKey` or `Digest`, since the algorithm would've already been validated by `normalizeAlgorithm` in all cases. But it doesn't hurt leaving that in, I suppose.